### DISCUSSION TITLE
Google Pub/Sub scaler not working properly

### DISCUSSION THREAD

**Discussion by led94** [kedacore/keda#4113] [Q&A / Need Help] [answered]
I've been having an issue following the sample for Pub/Sub scaling, https://github.com/kedacore/sample-go-gcppubsub, basically once I setup everything and try testing it the scaler goes to one pod but does not go beyond no matter how many messages are sent to the Topic and after a few minutes it goes back to 0, checking the logs it gives me the following error :

2022-12-09T03:16:02Z    ERROR   gcp_pub_sub_scaler      error getting Active Status     {"type": "ScaledObject", "namespace": "keda-pubsub-test", "name": "pubsub-scaledobject", "error": "rpc error: code = PermissionDenied desc = Permission monitoring.timeSeries.list denied (or the resource may not exist)."}

I gave the service account the monitoring viewer , the pubsub subscriber and the monitoring timeSeries list roles but no luck. 

Cluster version: 1.21.14-gke.4300
Keda version:  2.8.1

The code I am using is basically the same from the example with just a minor change, I am using mode and value instead of subscriptionsize as recommended.  I found this existing issue, https://github.com/kedacore/keda/issues/2208, but was not really helpful. 

Hope someone can help me sort this out.

**Comment by JorTurFer** [comment, marked as answer]
Hi @led94 
Sorry, I missed this discussion.
This seems like a problem in the metrics server, maybe requesting the metrics to GCP or maybe between k8s api-server and the metrics server. 
Could you share metrics server logs? In parallel, please check if the metric server is recogniced. You can do it with 
```bash
kubectl get apiservices v1beta1.external.metrics.k8s.io -o yaml
```
What do you see as the output of the command above?

**Reply by led94** [reply]
Hi @JorTurFer, 

Here are the logs from the metrics server:

$ kubectl logs keda-metrics-apiserver-5df7c98794-j7qk5 -n keda
I1207 17:37:02.381830       1 main.go:169] keda_metrics_adapter "msg"="KEDA Version: 2.8.1" 
I1207 17:37:02.382004       1 main.go:170] keda_metrics_adapter "msg"="KEDA Commit: 12783c1340a13ba6776d1d6a64127d88e7828aab" 
I1207 17:37:02.382035       1 main.go:171] keda_metrics_adapter "msg"="Go Version: go1.17.13"
I1207 17:37:02.382058       1 main.go:172] keda_metrics_adapter "msg"="Go OS/Arch: linux/amd64"
I1207 17:37:03.433723       1 request.go:601] Waited for 1.032959626s due to client-side throttling, not priority and fairness, request: GET:https://198.10.128.1:443/apis/rbac.authorization.k8s.io/v1beta1?timeout=32s
I1207 17:37:03.991274       1 logr.go:261] keda_metrics_adapter/controller-runtime/metrics "msg"="Metrics server is starting to listen" "addr"=":8080"
I1207 17:37:03.994140       1 provider.go:65] keda_metrics_adapter/provider "msg"="starting"
I1207 17:37:03.994179       1 main.go:236] keda_metrics_adapter "msg"="starting adapter..."
2022/12/07 17:37:03 Starting metrics server at :9022
I1207 17:37:03.997111       1 internal.go:362] keda_metrics_adapter "msg"="Starting server" "addr"={"IP":"::","Port":8080,"Zone":""} "kind"="metrics" "path"="/metrics"       
I1207 17:37:03.997457       1 controller.go:185] keda_metrics_adapter "msg"="Starting EventSource" "controller"="scaledobject" "controllerGroup"="keda.sh" "controllerKind"="ScaledObject" "source"="kind source: *v1alpha1.ScaledObject"
I1207 17:37:03.997521       1 controller.go:185] keda_metrics_adapter "msg"="Starting EventSource" "controller"="scaledobject" "controllerGroup"="keda.sh" "controllerKind"="ScaledObject" "source"="kind source: *v1alpha1.ScaledObject"
I1207 17:37:03.997541       1 controller.go:193] keda_metrics_adapter "msg"="Starting Controller" "controller"="scaledobject" "controllerGroup"="keda.sh" "controllerKind"="ScaledObject"
I1207 17:37:04.105759       1 controller.go:227] keda_metrics_adapter "msg"="Starting workers" "controller"="scaledobject" "controllerGroup"="keda.sh" "controllerKind"="ScaledObject" "worker count"=1
I1207 17:37:04.462108       1 serving.go:342] Generated self-signed cert (apiserver.local.config/certificates/apiserver.crt, apiserver.local.config/certificates/apiserver.key)
I1207 17:37:05.166191       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I1207 17:37:05.166924       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"        
I1207 17:37:05.166946       1 shared_informer.go:255] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1207 17:37:05.167448       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I1207 17:37:05.167491       1 shared_informer.go:255] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file 
I1207 17:37:05.166934       1 shared_informer.go:255] Waiting for caches to sync for RequestHeaderAuthRequestController
I1207 17:37:05.166820       1 secure_serving.go:210] Serving securely on [::]:6443
I1207 17:37:05.166870       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I1207 17:37:05.166899       1 dynamic_serving_content.go:132] "Starting controller" name="serving-cert::apiserver.local.config/certificates/apiserver.crt::apiserver.local.config/certificates/apiserver.key"
I1207 17:37:05.268064       1 shared_informer.go:262] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1207 17:37:05.268060       1 shared_informer.go:262] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1207 17:37:05.268894       1 shared_informer.go:262] Caches are synced for RequestHeaderAuthRequestController

While for the metric server, seems it is missing

$ kubectl get apiservices v1beta1.external.metrics.k8s.io -o yaml
Error from server (NotFound): apiservices.apiregistration.k8s.io "v1beta1.external.metrics.k8s.io" not found

**Reply by JorTurFer** [reply]
That's the problem. `v1beta1.external.metrics.k8s.io` is the metrics server registration and it says to the cluster where the metrics server is. How do you deploy KEDA? using helm? did you see any issue during the installation?

**Reply by led94** [reply]
We have had some issues with the installation, we had to copy the image to our repository and launch it from there, seems a network configuration on the cluster does not allow us to download the image from the repository and have not found how to solve it, so we make a manual deploy with the image we added to our own repository. I was not who installed it so I am going to cleanup the resources and try to install it and see if there are any errors

**Reply by JorTurFer** [reply]
It looks like there was any issue during the installation because as I said `v1beta1.external.metrics.k8s.io` must be there for KEDA, let me know once you have more info and we can continue :)

**Reply by led94** [reply]
Hi again @JorTurFer, 

I was able to redeploy the installation and now I guess I have a new little error, when running kubectl get apiservices v1beta1.external.metrics.k8s.io -o yaml I get the following:

status:
  conditions:
  - lastTransitionTime: "2023-01-24T14:51:03Z"
    message: 'failing or missing response from https://10.196.8.163:6443/apis/external.metrics.k8s.io/v1beta1:
      Get "https://10.196.8.163:6443/apis/external.metrics.k8s.io/v1beta1": context
      deadline exceeded (Client.Timeout exceeded while awaiting headers)'
    reason: FailedDiscoveryCheck
    status: "False"
    type: Available
    
I am following the troubleshooting guide on https://keda.sh/docs/2.9/troubleshooting/, going trough the firewall rule step, will let you know how it goes

**Reply by led94** [reply]
All worked now! Thank you so much @JorTurFer !!!
